### PR TITLE
[feature] #2712: Config proptests

### LINKED ISSUE #2712: Write proptests to check `Proxy` invariants
Probably something along [these](https://altsysrq.github.io/proptest-book/proptest/tutorial/recursive.html) lines.

### PR DESCRIPTION
**Pull request by ilchu**
### Description of the Change
* Added property testing to the Iroha and Iroha client `ConfigurationProxy` structs
* Expanded default `config.json` to allow them to be built from proxies immediately upon reading
* Updates some docs, error messages, trait derives 

### Issue
Closes #2712.

### Benefits
More confidence in current proxy implementation's robustness.

### Possible Drawbacks
I've struggled a bit with a generation strategy that would be compact yet showing the intended invariant, so I hope nothing's missing.

### Usage Examples or Tests 
`cargo test --package iroha_config --lib -- iroha::tests client::tests --nocapture`

### CODE REVIEW

File: primitives/src/small.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub const SMALL_SIZE: usize = 8_usize;
 mod small_string {
     use super::*;
 
-    #[derive(Debug, Clone, derive_more::Display, Deserialize, Serialize)]
+    #[derive(Debug, Clone, derive_more::Display, Deserialize, Serialize, PartialEq, Eq)]
```
Comment (appetrosyan) [inline primitives/src/small.rs:24]: In general, the more we derive the better.

File: config/src/client.rs [hunk 2]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::torii::{uri::DEFAULT_API_URL, DEFAULT_TORII_TELEMETRY_URL};
+
+    const CONFIGURATION_PATH: &str = "../configs/client_cli/config.json";
+
+    #[allow(clippy::expect_used)]
+    fn placeholder_keypair() -> KeyPair {
+        let public_key = "ed01207233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
```
Comment (appetrosyan) [inline config/src/client.rs:122]: Style choice, but I'd prefer this be derived from the private key.

File: config/src/client.rs [hunk 3]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+
+        KeyPair::new(public_key, private_key).expect("Key pair mismatch")
+    }
+
+    #[allow(clippy::expect_used)]
+    fn placeholder_account() -> <Account as Identifiable>::Id {
+        AccountId::from_str("alice@wonderland").expect("Account ID not valid")
+    }
+
+    fn arb_keys() -> BoxedStrategy<(Option<PublicKey>, Option<PrivateKey>)> {
```
Comment (appetrosyan) [inline config/src/client.rs:138]: I'm not 100% sure about this. Most keys are completely random so a proptest does literally nothing. Could we run tests with edge cases (one missing key, other missing key, both present but don't match), and leave this out of the proptest?

File: config/src/client.rs [hunk 4]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+            .boxed()
+    }
+
+    prop_compose! {
+        fn arb_proxy()
+            (
+                (public_key, private_key) in arb_keys(),
+                account_id in prop::option::of(Just(placeholder_account())),
+                basic_auth in prop::option::of(Just(Some(BasicAuth{web_login:WebLogin(SmallStr::from_str("mad_hatter")),password:SmallStr::from_str("ilovetea")}))),
+                torii_api_url in prop::option::of(Just(SmallStr::from_str(DEFAULT_API_URL))),
```
Comment (appetrosyan) [inline config/src/client.rs:153]: Actually here it'd be nice if we could also test different API URLs (e.g. one that's malformed, `http`, `ftp`, `ws` and so on. Things which are obviously wrong should be caught early, so someone who put in an extra `https` knows what went wrong before the peer is up.

File: config/src/client.rs [hunk 5]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+                torii_telemetry_url in prop::option::of(Just(SmallStr::from_str(DEFAULT_TORII_TELEMETRY_URL))),
```
Comment (appetrosyan) [inline config/src/client.rs:154]: Same here, except we know that this also has to provide a port, because the one case where the port was optional is the previous case. If the URLs are not identical, this warrants a warning in the logs, because it **might be** but isn't necessarily a mistake.

File: config/src/client.rs [hunk 6]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+        )
+            .boxed()
+    }
+
+    prop_compose! {
+        fn arb_proxy()
+            (
+                (public_key, private_key) in arb_keys(),
+                account_id in prop::option::of(Just(placeholder_account())),
+                basic_auth in prop::option::of(Just(Some(BasicAuth{web_login:WebLogin(SmallStr::from_str("mad_hatter")),password:SmallStr::from_str("ilovetea")}))),
```
Comment (appetrosyan) [inline config/src/client.rs:152]: I wouldn't test this at all. Basic auth is completely optional.

File: config/src/client.rs [hunk 7]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+                torii_api_url in prop::option::of(Just(SmallStr::from_str(DEFAULT_API_URL))),
+                torii_telemetry_url in prop::option::of(Just(SmallStr::from_str(DEFAULT_TORII_TELEMETRY_URL))),
+                transaction_time_to_live_ms in prop::option::of(Just(DEFAULT_TRANSACTION_TIME_TO_LIVE_MS)),
```
Comment (appetrosyan) [inline config/src/client.rs:155]: Probably going to be fine if we test edge-cases only. Although I'd probably throw a warning when the TTL is like a few milliseconds.

File: config/src/client.rs [hunk 8]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+                transaction_status_timeout_ms in prop::option::of(Just(DEFAULT_TRANSACTION_STATUS_TIMEOUT_MS)),
```
Comment (appetrosyan) [inline config/src/client.rs:156]: Check if this is shorter than TTL. Again, can be done via edge cases.

File: config/src/client.rs [hunk 9]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+                transaction_limits in prop::option::of(Just(TransactionLimits {
```
Comment (appetrosyan) [inline config/src/client.rs:157]: Probably warrant a warning if below certain threshold. `1` instruction per transaction is probably not a problem, but WASM has a certain minimum size. If the max is set too low, we should notify the user that they're doing something silly.

File: config/src/genesis.rs [hunk 10]
```diff
@@ -43,3 +43,48 @@ impl Default for ConfigurationProxy {
...
+            prop::option::of(Just(None)),
+        )
+            .boxed()
+    }
+
+    prop_compose! {
+        pub fn arb_proxy()
+            (
+                (account_public_key, account_private_key) in arb_keys(),
+                wait_for_peers_retry_count_limit in prop::option::of(Just(DEFAULT_WAIT_FOR_PEERS_RETRY_COUNT_LIMIT)),
```
Comment (appetrosyan) [inline config/src/genesis.rs:82]: Just like above, I'd consider adding warnings (but in a separate ticket).

File: config/src/client.rs (hunk 2 above)
Comment (Erigara) [reply config/src/client.rs:122]: This would require adding `From<PrivateKey>`, we have this [discussion](https://github.com/hyperledger/iroha/pull/2704#discussion_r970818247) in one of previous PR.
Maybe this PR is good opportunity to add this.

File: config/src/client.rs [hunk 11]
```diff
@@ -106,3 +106,75 @@ impl Default for ConfigurationProxy {
...
+    use iroha_config_base::proxy::{Builder, LoadFromDisk};
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::torii::{uri::DEFAULT_API_URL, DEFAULT_TORII_TELEMETRY_URL};
+
+    const CONFIGURATION_PATH: &str = "../configs/client_cli/config_test.json";
+
+    #[allow(clippy::expect_used)]
+    fn placeholder_keypair() -> KeyPair {
```
Comment (Erigara) [inline config/src/client.rs:121]: Another option would be to generate `KeyPair` with seed and seed would be determined by proptest.

File: config/src/client.rs [hunk 12]
```diff
@@ -60,7 +60,7 @@ pub struct BasicAuth {
 }
 
 /// `Configuration` provides an ability to define client parameters such as `TORII_URL`.
-#[derive(Debug, Clone, Deserialize, Serialize, Proxy, LoadFromEnv, Documented)]
+#[derive(Debug, Clone, Deserialize, Serialize, Proxy, LoadFromEnv, Documented, PartialEq, Eq)]
 #[serde(rename_all = "UPPERCASE")]
 #[config(env_prefix = "IROHA_")]
 pub struct Configuration {
```
Comment (Erigara) [inline config/src/client.rs:66]: I think we could add here `finalize` method as well and in finalize we would check configuration invaraints, log warnings and so on.

**Review by Erigara — COMMENTED** [review]
I like the first steps in direction of property based testing. 
IMO if done right (especially something like [stateful-testsing](https://hypothesis.readthedocs.io/en/latest/stateful.html)) it could help us to find bugs that would be hard to find otherwise.
John Hughes has great talks on this topic.

File: config/src/torii.rs [hunk 13]
```diff
@@ -76,3 +76,24 @@ pub mod uri {
...
 }
+
+#[cfg(test)]
+pub mod tests {
+    use proptest::prelude::*;
+
+    use super::*;
+
+    prop_compose! {
+        pub fn arb_proxy()
```
[the remark this answers is not in the document]
Comment (ilchu) [reply config/src/torii.rs:87]: Arbitrary (as in randomly generated) proxy. Just following the naming convention from the `proptest` docs.

File: config/src/sumeragi.rs [hunk 14]
```diff
@@ -90,7 +90,15 @@ impl ConfigurationProxy {
         self.trusted_peers = Some(TrustedPeers { peers });
     }
 }
+
 impl Configuration {
+    /// Set `trusted_peers` configuration parameter. Will overwrite
+    /// existing `trusted_peers` but does not check for duplication.
```
[the remark this answers is not in the document]
Comment (ilchu) [reply config/src/sumeragi.rs:96]: It's an old method that's used nowhere, so maybe it would be more correct to just delete it? But what the doc meant I suppose is that it doesn't warn the user if some of the provided peers were identical.

[the remark this answers is not in the document]
Comment (Erigara) [reply config/src/sumeragi.rs:96]: I don't mind deleting this method if it doesn't used anywhere.

File: config/src/client.rs (hunk 6 above)
Comment (ilchu) [reply config/src/client.rs:152]: I wouldn't say it's actually being tested, as that default value is wrapped in an option, and only the option is being flipped. But yeah, `None` looks and reads better.

File: config/src/client.rs (hunk 11 above)
Comment (ilchu) [reply config/src/client.rs:121]: I'll choose generating from a private key hex, as having that additional `From` impl might be beneficial for other parts of the code in the future.

**Comment by appetrosyan** [comment]
Just a vocabulary note. The word `finalize` has the `-ize` ending, the spelling of which cannot be agreed upon. If we write `ise` we'll annoy all American contributors, while writing `-ize` annoys British English speakers. 

Can we come up with a different word? For example: `finish`, `commit`, `complete`, hell even `consummate`?

File: config/src/genesis.rs (hunk 10 above)
Comment (ilchu) [reply config/src/genesis.rs:82]: Created https://github.com/hyperledger/iroha/issues/2801

File: config/src/client.rs (hunk 11 above)
Comment (ilchu) [reply config/src/client.rs:121]: Update: I naively thought that any randomly generated 512-bit hex would properly decompress into a key, which turned out not to be the case. Thus going for generation from seed, but still adding that `From` impl.

File: config/src/client.rs [hunk 15]
```diff
@@ -106,3 +106,166 @@ impl Default for ConfigurationProxy {
...
+        }
+        if let Some(tx_limits) = self.transaction_limits {
+            // 1 KiB
+            if tx_limits.max_wasm_size_bytes < 2_u64.pow(10) {
+                eyre::bail!(ConfigError::ProxyBuildError("`TRANSACTION_LIMITS` parameter's `max_wasm_size` field really small, consider making it bigger".to_owned()));
+            }
+        }
+        if let Some(api_url) = &self.torii_api_url {
+            let api_url = api_url.clone().to_string();
+            let split_api_url = api_url.split("://").collect::<Vec<_>>();
```
Comment (Erigara) [inline config/src/client.rs:144]: Could be done without allocation, but i wouldn't insist on this. Since we execute this method only once and i remember you suggestion about strong typing.

File: config/src/iroha.rs [hunk 16]
```diff
@@ -151,79 +148,101 @@ impl ConfigurationProxy {
...
 
     #[test]
     fn parse_example_json() -> Result<()> {
         let cfg_proxy = ConfigurationProxy::from_path(CONFIGURATION_PATH)
             .wrap_err("Failed to read configuration from example config")?;
-        assert_eq!(
-            "./storage",
-            cfg_proxy.kura.unwrap().block_store_path.unwrap()
-        );
+        assert_eq!("127.0.0.1:1337", cfg_proxy.torii.unwrap().p2p_addr.unwrap());
```
Comment (Erigara) [inline config/src/iroha.rs:230]: Why do we switch this line back and forth?

File: config/src/iroha.rs [hunk 17]
```diff
@@ -151,79 +148,101 @@ impl ConfigurationProxy {
...
+
     use super::*;
     use crate::{base::proxy::LoadFromDisk, sumeragi::TrustedPeers};
 
-    const CONFIGURATION_PATH: &str = "../configs/peer/config.json";
+    const CONFIGURATION_PATH: &str = "../configs/peer/config_test.json";
+
+    /// Key-pair used for proptests generation
+    #[allow(clippy::expect_used)]
+    pub fn placeholder_keypair() -> KeyPair {
```
Comment (Erigara) [inline config/src/iroha.rs:173]: Why not generate keys here as well?

File: config/base/tests/simple.rs [hunk 18]
```diff
@@ -6,7 +6,7 @@ use iroha_config_base::{
 };
 use serde::{Deserialize, Serialize};
 
-#[derive(Clone, Debug, Deserialize, Serialize, Documented, LoadFromEnv)]
+#[derive(Clone, Debug, Deserialize, Serialize, Documented, LoadFromEnv, PartialEq, Eq)]
```
Comment (appetrosyan) [inline config/base/tests/simple.rs:9]: In general, if some trait can be derived, it should be derived in library crates. Not so in binary crates.

File: config/src/client.rs (hunk 12 above)
Comment (appetrosyan) [reply config/src/client.rs:66]: I propose a different naming convention, one that avoids the `ize`/`ise` suffices.

File: config/src/client.rs (hunk 15 above)
Comment (appetrosyan) [reply config/src/client.rs:144]: I prefer it this way. The allocation is only a problem when you do several times in a row. Almost everything related to configuration is run once, so flexibility is more important.

File: crypto/src/lib.rs [hunk 19]
```diff
@@ -439,6 +439,27 @@ impl From<Multihash> for PublicKey {
...
+        let key_gen_option = Some(UrsaKeyGenOption::FromSecretKey(UrsaPrivateKey(
+            private_key.payload,
+        )));
+        let (mut public_key, _) = match digest_function {
+            Algorithm::Ed25519 => Ed25519Sha512.keypair(key_gen_option),
+            Algorithm::Secp256k1 => EcdsaSecp256k1Sha256::new().keypair(key_gen_option),
+            Algorithm::BlsNormal => BlsNormal::new().keypair(key_gen_option),
+            Algorithm::BlsSmall => BlsSmall::new().keypair(key_gen_option),
+        }
+        .expect("can't fail for valid `PrivateKey`");
```
Comment (appetrosyan) [inline crypto/src/lib.rs:455]: We should actually fix this upstream in hyperledger/ursa

File: configs/peer/config_test.json [hunk 20]
```diff
@@ -0,0 +1,115 @@
+{
```
Comment (appetrosyan) [inline configs/peer/config_test.json:1]: If you need this file, it needs to be placed elsewhere.

Comment (appetrosyan) [reply configs/peer/config_test.json:1]: Given that this is used for testing of the `config` crate, I'd suggest putting it there locally.

**Review by appetrosyan — CHANGES_REQUESTED** [review]

File: config/src/iroha.rs (hunk 16 above)
Comment (appetrosyan) [reply config/src/iroha.rs:230]: Formatting. Small changes lead to it being either written on separate lines, or on the same line.

File: config/src/iroha.rs (hunk 17 above)
Comment (appetrosyan) [reply config/src/iroha.rs:173]: This is going to add stochasticity to the (already flaky) tests. We need to remove as many variables as we can, unless they are part of the proptest. In that case, the keys will need to be generated from a seed that's provide by the proptest framework.

File: crypto/src/lib.rs (hunk 19 above)
Comment (ilchu) [reply crypto/src/lib.rs:455]: If you mean filing a PR with this impl into their repo, then I'm a bit skeptical about the speed of their PR acceptance process, but I could try to.

File: config/src/client.rs [hunk 21]
```diff
@@ -106,3 +106,166 @@ impl Default for ConfigurationProxy {
         }
     }
 }
+
+impl ConfigurationProxy {
+    /// Finalise Iroha config proxy by instantiating mutually equivalent fields
```
[the remark this answers is not in the document]
Comment (ilchu) [reply config/src/client.rs:111]: Missed updating it, thanks!

File: config/src/iroha.rs (hunk 16 above)
Comment (ilchu) [reply config/src/iroha.rs:230]: Actually it's not about formatting here, but due to me mindlessly rewriting back changes introduced by [this](https://github.com/hyperledger/iroha/pull/2782#discussion_r980902572). Will revert it back.

File: crypto/src/lib.rs (hunk 19 above)
Comment (ilchu) [reply crypto/src/lib.rs:455]: On further inspection, ursa has nothing but opaque bytearrays in place of private / public keys with no associated digests, and changing the structs to a product type would break FFI, macros, and who knows what else. So I guess the cost is not worth it.

File: config/iroha_test_config.json [hunk 22]
```diff
@@ -0,0 +1,115 @@
+{
```
Comment (appetrosyan) [inline config/iroha_test_config.json:1]: Much better.

**Review by appetrosyan — DISMISSED** [review]

**Review by Erigara — APPROVED** [review]

File: config/src/client.rs [hunk 23]
```diff
@@ -106,3 +106,165 @@ impl Default for ConfigurationProxy {
         }
     }
 }
+
+impl ConfigurationProxy {
+    /// Finalise Iroha client config proxy by checking that certain fields identify reasonable limits or
+    /// or are well formatted.
```
Comment (outoftardis) [inline config/src/client.rs:112]:
```suggestion
    /// are well formatted.
```

File: config/src/client.rs [hunk 24]
```diff
@@ -106,3 +106,165 @@ impl Default for ConfigurationProxy {
...
+            }
+        }
+        Ok(())
+    }
+
+    /// The wrapper around the client `ConfigurationProxy` that performs
+    /// finalisation prior to building `Configuration`. Just like
+    /// Iroha peer config, its `<Self as iroha_config_base::proxy::Builder>::build()`
+    /// method should never be used directly, as only this wrapper ensures final
+    /// coherence and issues important warnings.
```
Comment (outoftardis) [inline config/src/client.rs:180]: maybe
```suggestion
    /// coherence and warns if there are any issues.
```

**Review by appetrosyan — DISMISSED** [review]

**Review by SamHSmith — APPROVED** [review]
LGTM

**Review by Erigara — APPROVED** [review]
